tree/tree/test/TChainParsing.cxx

```diff
@@ -184,6 +185,92 @@ TEST(TChainParsing, GlobbingWithNonExistingDir)
+   return chainFileNames;
+}
+
+TEST(TChainParsing, RecursiveGlob)
+{
+   // Need to change working directory to ensure "*" only adds files we create here.
+   auto testDir = "testdir";
+   ASSERT_EQ(gSystem->mkdir(testDir), 0);
+   gSystem->ChangeDirectory(testDir);
+
+   const auto *cwd = gSystem->WorkingDirectory();
+
+   ASSERT_EQ(gSystem->mkdir("testglob"), 0);
+   ASSERT_EQ(gSystem->mkdir("testglob/subdir1"), 0);
+   ASSERT_EQ(gSystem->mkdir("testglob/subdir1/subdir11"), 0);
+   ASSERT_EQ(gSystem->mkdir("testglob/subdir2"), 0);
+   ASSERT_EQ(gSystem->mkdir("testglob/subdir2/subdir21"), 0);
+   ASSERT_EQ(gSystem->mkdir("testglob/subdir3"), 0);
+
+   // Unsorted to also test sorting the final list of globbed files.
+   std::vector<std::string> fileNames{"0a.root",
+                                      "testglob/subdir1/1a.root",
+                                      "testglob/subdir1/1b.root",
+                                      "testglob/subdir3/3a.root",
+                                      "testglob/subdir2/subdir21/21a.root",
+                                      "testglob/subdir2/subdir21/21b.root",
+                                      "testglob/subdir1/subdir11/11b.root",
+                                      "testglob/subdir1/subdir11/11a.root"};
+   for (const auto &fileName : fileNames) {
+      FillTree(fileName.c_str(), "events");
+   }
+
+   TChain nodir, none, nested, globDir;
+
+   nodir.Add("*");
+   const auto *nodirFiles = nodir.GetListOfFiles();
+   ASSERT_TRUE(nodirFiles);
+   EXPECT_EQ(nodirFiles->GetEntries(), 1);
+   auto expectedFileNameNodir = gSystem->ConcatFileName(cwd, "0a.root");
+   EXPECT_STREQ(nodirFiles->At(0)->GetTitle(), gSystem->UnixPathName(expectedFileNameNodir));
+   delete[] expectedFileNameNodir;
+
+   none.Add("testglob/*");
+   const auto *noneChainFiles = none.GetListOfFiles();
+   ASSERT_TRUE(noneChainFiles);
+   EXPECT_EQ(noneChainFiles->GetEntries(), 0);
+
+   nested.Add("testglob/*/*/*.root");
+   const auto *nestedChainFiles = nested.GetListOfFiles();
+   ASSERT_TRUE(nestedChainFiles);
+   EXPECT_EQ(nestedChainFiles->GetEntries(), 4);
+   std::vector<std::string> expectedFileNamesNested{
+      "testglob/subdir1/subdir11/11a.root", "testglob/subdir1/subdir11/11b.root", "testglob/subdir2/subdir21/21a.root",
+      "testglob/subdir2/subdir21/21b.root"};
+   auto nestedChainFileNames = GetFileNamesVec(nestedChainFiles);
+   EXPECT_VEC_EQ(nestedChainFileNames, expectedFileNamesNested);
+
+   globDir.Add("*/subdir[0-9]/*");
+   const auto *globDirChainFiles = globDir.GetListOfFiles();
+   ASSERT_TRUE(globDirChainFiles);
+   EXPECT_EQ(globDirChainFiles->GetEntries(), 3);
+   std::vector<char *> expectedFileNamesGlobDir{gSystem->ConcatFileName(cwd, "testglob/subdir1/1a.root"),
+                                                gSystem->ConcatFileName(cwd, "testglob/subdir1/1b.root"),
+                                                gSystem->ConcatFileName(cwd, "testglob/subdir3/3a.root")};
```

Comment:
```c++
   std::vector<std::string> expectedFileNamesGlobDir{concatFileName(cwd, "testglob/subdir1/1a.root"),
                                                     concatFileName(cwd, "testglob/subdir1/1b.root"),
                                                     concatFileName(cwd, "testglob/subdir3/3a.root")};
```
Carrying around these weird owning C-style strings is not great.

I would introduce a helper function to convert them early:
```c++
std::string concatFileName(const char *dir, const char *name)
{
   return std::unique_ptr<char[]>{gSystem->ConcatFileName(dir, name)}.get();
}
```